AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Core/DataStoreList.swift

```diff
@@ -0,0 +1,172 @@
+//
+// Copyright 2018-2020 Amazon.com,
+// Inc. or its affiliates. All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+import Amplify
+import Combine
+
+/// `DataStoreList<ModelType>` is a DataStore-aware custom `Collection` that is capable of loading
+/// records from the `DataStore` on-demand. This is specially useful when dealing with
+/// Model associations that need to be lazy loaded.
+///
+/// When using `DataStore.query(_ modelType:)` some models might contain associations
+/// with other models and those aren't fetched automatically. This collection keeps track
+/// of the associated `id` and `field` and fetches the associated data on demand.
+public class DataStoreList<ModelType: Model>: List<ModelType>, ModelListDecoder {
+
+    /// If the list represents an association between two models, the `associatedId` will
+    /// hold the information necessary to query the associated elements (e.g. comments of a post)
+    internal var associatedId: Model.Identifier?
+
+    /// The associatedField represents the field to which the owner of the `List` is linked to.
+    /// For example, if `Post.comments` is associated with `Comment.post` the `List<Comment>`
+    /// of `Post` will have a reference to the `post` field in `Comment`.
+    internal var associatedField: ModelField?
+
+    internal var limit: Int = 100
+
+    /// The current state of lazily loaded list
+    internal var state: LoadState = .pending
+
+    // MARK: - Initializers
+
+    public override convenience init(_ elements: Elements) {
+        self.init(elements, associatedId: nil, associatedField: nil)
+        self.state = .loaded
+    }
+
+    init(_ elements: Elements,
+         associatedId: Model.Identifier? = nil,
+         associatedField: ModelField? = nil) {
+        super.init(elements)
+        self.associatedId = associatedId
+        self.associatedField = associatedField
+    }
+
+    // MARK: - ExpressibleByArrayLiteral
+
+    required convenience public init(arrayLiteral elements: Element...) {
+        self.init(elements)
+    }
+
+    // MARK: - Collection conformance
+
+    public override var startIndex: Index {
```

Comment:
Here and anywhere we do a `loadIfNeeded`: note that this method implicitly lazy loads, so it isn't O(1) unless the collection has already been `load`ed